server/EXCEPTIONS.md

```diff
@@ -0,0 +1,192 @@
+# Exception handling in our backend
+
+## Quickstart
+
+To generate errors in the backend that will be passed to the
+frontend, throw a [Nest-defined HttpException](https://docs.nestjs.com/exception-filters#throwing-standard-exceptions), <u>_generally within a service_</u>:
+
+```js
+  throw new HttpException(responseBody, status)
+```
+
+Where...
+  - `status` is one of `HttpStatus` enum codes. E.g. `HttpStatus.UNAUTHORIZED`
+  - `responseBody` argument is an object with the following schema.
+    ```js
+       const responseBody = {
+        code: 'APPLICANT_PORTAL_ERROR_CODE',
+        message: 'Applicant Portal could not find coffee. CRM: CRM ran out of tea.',
+      }
+    ```
+    - The `code` property is developer-defined, and should be a unique, concise,
+    descriptive identifier for the type of error. Users can provide this code to
+    help developers locate the code that yielded the error.
+    - The `message` property is a short, human-readable sentence firstly
+    describing the error for the benefit of end users. Secondly, it should contain
+    errors relayed from CRM, if available.
+
+To scope the error to a specific part of the code, use try-catch blocks. For example,
+the following try-catch scopes the thrown error to an attempt to find a project.
+
+```ts
+// services/my-project.service.ts
+public async getProject(){
+  try {
+    return myProject = this.get('dcp_project', ...);
+  } catch (e) {
+    throw new HttpException({
+      code: 'GET_PROJECTS',
```

Comment:
Not sure I agree with this style...

in HTTP, a status _code_ is something like "404 NOT FOUND" — it tries to explain the status.

In this case, the code describes what named object it comes from, including method. That's useful information, but I think it would go somewhere else, possible in `meta` or just added to the `detail` key

Reply:
Hmm. The JSON API do suggest the `code` is specific to the application.
```
code: an application-specific error code, expressed as a string value.
```

> code is similar to title in that both identify a general type of problem (unlike detail, which is specific to the particular instance of the problem), but dealing with code is easier programatically,

https://jsonapi.org/examples/#error-objects-error-codes


Their examples also suggest the same:

![image](https://user-images.githubusercontent.com/3311663/89061497-a3902f00-d319-11ea-9db6-972248c7b58c.png)

Reply:
Glad we're leaning more on JSON:API. I don't see how my comment conflicts with the language you're quoting. We want  to point to a "general type of problem", not scope the general problem in the code identifier with the named subclass itself.

Reply:
Ohh, I thought you were suggesting the code should be a slightly longer version of the `status` property.